flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java

```diff
@@ -650,6 +655,66 @@ public void testClusterDoesNOTShutdownWhenApplicationStatusUknown() throws Excep
         assertEquals(exception.getStatus(), ApplicationStatus.UNKNOWN);
     }
 
+    @Test
+    public void testDuplicateJobSubmissionWithTerminatedJobId() throws Throwable {
+        final JobID testJobID = new JobID(0, 2);
+        final Configuration configurationUnderTest = getConfiguration();
+        configurationUnderTest.set(
+                PipelineOptionsInternal.PIPELINE_FIXED_JOB_ID, testJobID.toHexString());
+        configurationUnderTest.set(
+                HighAvailabilityOptions.HA_MODE, HighAvailabilityMode.ZOOKEEPER.name());
+        final TestingDispatcherGateway.Builder dispatcherBuilder =
+                new TestingDispatcherGateway.Builder()
+                        .setSubmitFunction(
+                                jobGraph -> {
+                                    final CompletableFuture<Acknowledge> submit =
+                                            new CompletableFuture<>();
+                                    submit.completeExceptionally(
+                                            new DuplicateJobSubmissionException(testJobID, true));
+                                    return submit;
+                                })
+                        .setRequestJobStatusFunction(
+                                jobId -> CompletableFuture.completedFuture(JobStatus.FINISHED))
+                        .setRequestJobResultFunction(
+                                jobId ->
+                                        CompletableFuture.completedFuture(
+                                                createSuccessfulJobResult(jobId)));
```

Comment:
I am not sure whether this will work in all cases. I actually think that Flink loses the job result in case of a `JobManager` process failover. Hence, I would expect that the application mode will fail with a `FlinkJobNotFoundException` once it tries to query the result for the already finished job.

Reply:
Fixed. ApplicationDispatcherBootstrap now returns "unknown" JobResult in case job result can no longer be retrieved after fail-over.